### PR TITLE
Add C++ ops to torchvision

### PR DESCRIPTION
**Pull request by fmassa**
This PR adds C++ / CUDA extensions to torchvision for `ROIAlign`, `ROIPool` and `nms`.

I'll still be doing some cleanups / linting, as well as moving the files from `layers` to `ops` folder, so this is not yet ready to be merged.

Thanks a lot to @varunagrawal who added a number of improvements on top of the `layers` branch, such as CPU support for ROIAlign and ROIPool backwards, as well as unit tests for those operations!

### CODE REVIEW

**Comment by varunagrawal** [comment]
@fmassa are you sure about removing those `+ 1` for area computations in NMS? Assuming we are working in the XYXY box format, a box `(0, 0, 15, 15)` is technically 16x16 in size.

**Comment by fmassa** [comment]
@varunagrawal it depends on how we consider what the coordinates of the bounding box represent and if cropping is inclusive or exclusive on the boundary.

If indexing is inclusive (like in matlab), then yes, the +1 is necessary. But if it's exclusive (like in Python), then `x[0:15]` has 15 elements.

This is up for discussion, but I believe that we should be consistent everywhere.

Tagging @rbgirshick if he has any comments if we should include the `+1` or not in the dimensions / area calculation of bounding boxes.

**Comment by varunagrawal** [comment]
When I looked through the implementation of IoU on maskrcnn-benchmark, the +1 is always included, so there's precedence that requires that.

More importantly, I don't think it is a indexing issue. For example, in COCO the boxes are XYWH e.g. `(0, 0, 16, 16)`, and to get the RoI, you are right that we would need to use `x[X:W, Y:W]` if it's exclusive lilke python.
However, as per the detection literature, the model gives us the bounding boxes in XYXY format, and to get the width and height of the bounding box correctly, we need to offset by `+1`. Using the example of `(0, 0, 16, 16)` in XYWH mode, the bottom right pixel in the RoI would be `(15, 15)`, giving us `(0, 0, 15, 15)` in XYXY mode, since if you count the number of pixels, 0-15 gives us 16 pixels which is the correct size.

File: torchvision/csrc/nms.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    const at::Tensor& dets,
+    const at::Tensor& scores,
+    const float threshold) {
+  if (dets.device().is_cuda()) {
+#ifdef WITH_CUDA
+    if (dets.numel() == 0) {
+      at::cuda::CUDAGuard device_guard(dets.device());
+      return at::empty({0}, dets.options().dtype(at::kLong));
+    }
+    auto b = at::cat({dets, scores.unsqueeze(1)}, 1);
```
Comment (varunagrawal) [inline torchvision/csrc/nms.h]: I've been wondering about this. It seems unnecessary to waste memory creating a new tensor here when we can easily just update the function signature to accept 3 arguments. Thoughts?

File: torchvision/csrc/cuda/nms_cuda.cu [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+  at::Tensor mask_cpu = mask.to(at::kCPU);
+  unsigned long long* mask_host = (unsigned long long*)mask_cpu.data<int64_t>();
+
+  std::vector<unsigned long long> remv(col_blocks);
+  memset(&remv[0], 0, sizeof(unsigned long long) * col_blocks);
+
+  at::Tensor keep =
+      at::empty({boxes_num}, boxes.options().dtype(at::kLong).device(at::kCPU));
+  int64_t* keep_out = keep.data<int64_t>();
```
Comment (varunagrawal) [inline torchvision/csrc/cuda/nms_cuda.cu]: If `keep` is changed to `kFloat`, this will need to change as well.

File: torchvision/csrc/nms.h (hunk 1 above)
Comment (fmassa) [reply torchvision/csrc/nms.h]: I totally agree, I just wanted to minimize the amount of work for now, but this is definitely something I'd like to have done

**Comment by fmassa** [comment]
@varunagrawal thanks for the comments. To address your points:

> When I looked through the implementation of IoU on maskrcnn-benchmark, the +1 is always included, so there's precedence that requires that.

`maskrcnn-benchmark` tried to be as close as possible to Detectron, to provide baselines in PyTorch for detection models. In particular, it was backwards-compatible with Detectron models. But Detectron itself carries a lot of baggage from a long time ago, and some of the decisions there I do not think apply anymore.

> More importantly, I don't think it is a indexing issue. For example, in COCO the boxes are XYWH e.g. (0, 0, 16, 16), and to get the RoI, you are right that we would need to use x[X:W, Y:W] if it's exclusive lilke python.
However, as per the detection literature, the model gives us the bounding boxes in XYXY format, and to get the width and height of the bounding box correctly, we need to offset by +1. Using the example of (0, 0, 16, 16) in XYWH mode, the bottom right pixel in the RoI would be (15, 15), giving us (0, 0, 15, 15) in XYXY mode, since if you count the number of pixels, 0-15 gives us 16 pixels which is the correct size.

I believe this is just a matter of convention. For example, in TF object detection library, [area and width/height do not include the +1](https://github.com/tensorflow/models/blob/master/research/object_detection/core/box_list_ops.py#L44-L74). This is particularly important because they use normalized coordinates (so that boxes are within 0-1). Plus, when converting [COCO dataset to their format, no +1 is added either](https://github.com/tensorflow/models/blob/1255d5b94dec8f1927e3d500db15ef2078196ecb/research/object_detection/dataset_tools/create_coco_tf_record.py#L138).
I tend to agree with their approach, which just means that in `x1y1x2y2`, the `x1y1` is inclusive and `x2y2` is exclusive.

If you see, the `xyxy<->xywh` conversion functions from `maskrcnn-benchmark` reproduce exactly those from Detectron. Still, when [going from `xywh` to `xyxy`](https://github.com/facebookresearch/maskrcnn-benchmark/blob/f917a555bc422ed5e06a402e739da0e21b00d0b5/maskrcnn_benchmark/structures/bounding_box.py#L85-L86), we remove `+1`. Then, when computing the area, the `+1` is added again. This is weird to me, and I'd like to remove those inconsistencies in torchvision (as we are not tied to matching exactly Detectron, as there is `maskrcnn-benchmark` for that).

There are plenty of inconsistencies in `maskrcnn-benchmark` that were carried from Detectron: [in `BoxList.crop`, the width/height computation does not add a `+1`](https://github.com/facebookresearch/maskrcnn-benchmark/blob/f917a555bc422ed5e06a402e739da0e21b00d0b5/maskrcnn_benchmark/structures/bounding_box.py#L174)

Let me know if you disagree

File: torchvision/csrc/nms.h (hunk 1 above)
Comment (varunagrawal) [reply torchvision/csrc/nms.h]: I'll add a PR on top of this branch to add that later today.

**Comment by varunagrawal** [comment]
> `maskrcnn-benchmark` tried to be as close as possible to Detectron, to provide baselines in PyTorch for detection models. In particular, it was backwards-compatible with Detectron models. But Detectron itself carries a lot of baggage from a long time ago, and some of the decisions there I do not think apply anymore.

I see. I guess having Ross' comments on this would be helpful then.

> I believe this is just a matter of convention. For example, in TF object detection library, [area and width/height do not include the +1](https://github.com/tensorflow/models/blob/master/research/object_detection/core/box_list_ops.py#L44-L74). This is particularly important because they use normalized coordinates (so that boxes are within 0-1). Plus, when converting [COCO dataset to their format, no +1 is added either](https://github.com/tensorflow/models/blob/1255d5b94dec8f1927e3d500db15ef2078196ecb/research/object_detection/dataset_tools/create_coco_tf_record.py#L138).
> I tend to agree with their approach, which just means that in `x1y1x2y2`, the `x1y1` is inclusive and `x2y2` is exclusive.

Okay this is what I was looking for. If `x2y2` is exclusive then that solves our problem. I was worried it was not, and we should make a note of it in the docs since quite a few people write code expecting it to be inclusive.

**Comment by rbgirshick** [comment]
The "+1" history is described here https://github.com/facebookresearch/Detectron/blob/master/detectron/utils/boxes.py#L28-L40.

I think this is baggage that should be shed. I believe the correct thing to do is consider a box as a geometric entity defined in terms of continuous coordinates. In this case width is simply `x2 - x1`. Discrete coordinates and indexing only needs to be considered when rasterizing a continuous entity, in which case a pixel should be thought of a point sample, not a little square (as described in http://alvyray.com/Memos/CG/Microsoft/6_pixel.pdf).

**Comment by lopuhin** [comment]
Thanks @fmassa , it would be great to have officially supported layers for object detection.
When doing some testing, I noticed a difference between GPU and CPU forward for roi pooling, with GPU being correct. I'll try to put a reproducible example.

**Comment by fmassa** [comment]
@lopuhin this might be the same issue as https://github.com/facebookresearch/maskrcnn-benchmark/issues/331

Also, I'm putting together a MWE of object detection and semantic segmentation, to be included in the `references` of torchvision

**Comment by lopuhin** [comment]
Thanks @fmassa , in my case the difference looks larger, e.g. 1.4492 vs 2.5916 for some values (and equal for other). It could be some bug on my side, will try to investigate soon.

**Comment by lopuhin** [comment]
@fmassa here is an example of roi pooling forward difference between GPU and CPU:
```
import torch
from torchvision.ops import roi_pool
x = torch.tensor(
    [[[[0.1767, 1.2851, 4.2325, 4.8645, 7.1496]],
      [[2.5916, 4.3361, 3.8143, 6.1329, 2.0230]],
      [[1.4492, 3.3384, 4.0816, 6.3116, 5.1068]]]],
    dtype=torch.float32)
rois = torch.tensor(
    [[0., 1., 0., 4., 0.],
     [0., 2., 0., 3., 0.],
     [0., 0., 0., 0., 0.],
     [0., 0., 0., 0., 0.],
     [0., 2., 0., 2., 0.]],
    dtype=torch.float32)
a = roi_pool(x, rois, (1, 2))
b = roi_pool(x.to('cuda'), rois.to('cuda'), (1, 2))
print(a - b.cpu())
```

which gives for me (blank lines removed for clarity):
```
tensor(
       [[[[0.0000, 0.0000]],
         [[0.0000, 1.0167]],
         [[0.2545, 0.8380]]],

        [[[0.0000, 0.0000]],
         [[0.4182, 0.0000]],
         [[0.1509, 0.0000]]],

        [[[0.0000, 0.0000]],
         [[0.0000, 0.0000]],
         [[1.1424, 1.1424]]],

        [[[0.0000, 0.0000]],
         [[0.0000, 0.0000]],
         [[1.1424, 1.1424]]],

        [[[0.0000, 0.0000]],
         [[0.4182, 0.4182]],
         [[0.1509, 0.1509]]]])
```

while I expected ``a - b`` to be close to zero here

**Comment by fmassa** [comment]
@lopuhin this looks indeed like a bug, I'll have a look before merging this

**Comment by soumith** [comment]
fyi, the CI is now good to go on this.

**Comment by fmassa** [comment]
@lopuhin I believe I've fixed the bug you pointed out. Thanks a lot for the repro!

**Comment by lopuhin** [comment]
Thank you very much for the fix and for the feature @fmassa , I confirm that it works as expected for me 👍

**Comment by TheCodez** [comment]
@fmassa can those ops be used from torchscript?

**Comment by fmassa** [comment]
@TheCodez probably not as of now, but I'll want to make them work in torchscript at some point in time

**Comment by TheCodez** [comment]
@fmass `nms` can probably just be added using the below code and loading the correct lib 
for the right target platform in `__init__.py`:
```cpp
static auto registry = torch::jit::RegisterOperators("vision::nms", &nms);
```

The other ops seem to be more involved.

**Comment by fmassa** [comment]
@TheCodez yes, and I'll follow the modifications from https://github.com/facebookresearch/maskrcnn-benchmark/pull/138 to make the operators here work with JIT, but that might require a few modifications elsewhere in PyTorch so that both can coexist, so I'm postponing this change.

**Comment by lopuhin** [comment]
FWIW I'm getting ``RuntimeError: CUDA error: an illegal memory access was encountered`` in ``ROIPool_backward_cuda``, I'll try to reduce this to a reproducible issue an post a separate issue

```
    @staticmethod
    @once_differentiable
    def backward(ctx, grad_output):
        rois, argmax = ctx.saved_tensors
        output_size = ctx.output_size
        spatial_scale = ctx.spatial_scale
        bs, ch, h, w = ctx.input_shape
        grad_input = _C.roi_pool_backward(
            grad_output, rois, argmax, spatial_scale,
>           output_size[0], output_size[1], bs, ch, h, w)
E       RuntimeError: CUDA error: an illegal memory access was encountered (ROIPool_backward_cuda at /tmp/pip-install-nz4z64qf/torchvision/torchvision/csrc/cuda/ROIPool_cuda.cu:240)
E       frame #0: std::function<std::string ()>::operator()() const + 0x11 (0x7ff03cd24441 in /usr/local/lib/python3.6/dist-packages/torch/lib/libc10.so)
E       frame #1: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x2a (0x7ff03cd23d7a in /usr/local/lib/python3.6/dist-packages/torch/lib/libc10.so)
E       frame #2: ROIPool_backward_cuda(at::Tensor const&, at::Tensor const&, at::Tensor const&, float, int, int, int, int, int, int) + 0xad3 (0x7ff029c08feb in /usr/local/lib/python3.6/dist-packages/torchvision/_C.cpython-36m-x86_64-linux-gnu.so)
E       frame #3: ROIPool_backward(at::Tensor const&, at::Tensor const&, at::Tensor const&, float, int, int, int, int, int, int) + 0x140 (0x7ff029bd3a60 in /usr/local/lib/python3.6/dist-packages/torchvision/_C.cpython-36m-x86_64-linux-gnu.so)
E       frame #4: <unknown function> + 0x22157 (0x7ff029be1157 in /usr/local/lib/python3.6/dist-packages/torchvision/_C.cpython-36m-x86_64-linux-gnu.so)
E       frame #5: <unknown function> + 0x1e461 (0x7ff029bdd461 in /usr/local/lib/python3.6/dist-packages/torchvision/_C.cpython-36m-x86_64-linux-gnu.so)
E       frame #6: /usr/bin/python3() [0x5030d5]
E       frame #7: _PyEval_EvalFrameDefault + 0x449 (0x506859 in /usr/bin/python3)
E       frame #8: /usr/bin/python3() [0x504c28]
E       frame #9: /usr/bin/python3() [0x58644b]
E       frame #10: PyObject_Call + 0x3e (0x59ebbe in /usr/bin/python3)
E       frame #11: _PyEval_EvalFrameDefault + 0x1807 (0x507c17 in /usr/bin/python3)
E       frame #12: /usr/bin/python3() [0x504c28]
E       frame #13: /usr/bin/python3() [0x58644b]
E       frame #14: PyObject_Call + 0x3e (0x59ebbe in /usr/bin/python3)
E       frame #15: _PyEval_EvalFrameDefault + 0x1807 (0x507c17 in /usr/bin/python3)
E       frame #16: /usr/bin/python3() [0x504c28]
E       frame #17: _PyFunction_FastCallDict + 0x2de (0x501b2e in /usr/bin/python3)
E       frame #18: /usr/bin/python3() [0x591461]
E       frame #19: PyObject_Call + 0x3e (0x59ebbe in /usr/bin/python3)
E       frame #20: torch::autograd::PyFunction::apply(std::vector<torch::autograd::Variable, std::allocator<torch::autograd::Variable> >&&) + 0x193 (0x7ff085755823 in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch_python.so)
E       frame #21: <unknown function> + 0x3108aa (0x7ff03b1ae8aa in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so.1)
E       frame #22: torch::autograd::Engine::evaluate_function(torch::autograd::FunctionTask&) + 0x385 (0x7ff03b1a7975 in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so.1)
E       frame #23: torch::autograd::Engine::thread_main(torch::autograd::GraphTask*) + 0xc0 (0x7ff03b1a9970 in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so.1)
E       frame #24: torch::autograd::Engine::thread_init(int) + 0x136 (0x7ff03b1a6d46 in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so.1)
E       frame #25: torch::autograd::python::PythonEngine::thread_init(int) + 0x2a (0x7ff0857502ea in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch_python.so)
E       frame #26: <unknown function> + 0xbd8f0 (0x7ff03c8278f0 in /usr/lib/x86_64-linux-gnu/libstdc++.so.6)
E       frame #27: <unknown function> + 0x76db (0x7ff09eb066db in /lib/x86_64-linux-gnu/libpthread.so.0)
E       frame #28: clone + 0x3f (0x7ff09ee3f88f in /lib/x86_64-linux-gnu/libc.so.6)

/usr/local/lib/python3.6/dist-packages/torchvision/ops/roi_pool.py:34: RuntimeError
```

**Comment by fmassa** [comment]
Yes, a repro on a new issue would be great!

**Comment by lopuhin** [comment]
Thanks @fmassa , so far it looks a bit more challenging to reproduce :)
Even adding ``time.sleep(1)`` before line 34 makes the issue much less likely to happen (without it, it always happens), and also adding ``grad_output = grad_output.contiguous()`` makes the issue go away. Also I tried on two machines so far and it happens only on one (although environments are slightly different). CuDNN is disabled and CUDA_LAUNCH_BLOCKING is set.
Still hope to reduce it, as it happens reliably.

**Comment by fmassa** [comment]
Don't we add a contiguous call to grad_output already?

**Comment by fmassa** [comment]
If not, we should

**Comment by lopuhin** [comment]
Oh interesting, I didn't know it's required, this is the place where it helps me: https://github.com/pytorch/vision/blob/37bd11dfc189adc28790ebaad63ddb3febc8c2ee/torchvision/ops/roi_pool.py#L32

**Comment by fmassa** [comment]
There is a call to contiguous in the cuda file, jest before the kernel launch, so the problem should be elsewhere

**Comment by lopuhin** [comment]
Right, found it here https://github.com/pytorch/vision/blob/dc3ac2909c2b9a5f6874ba1f1041ef49bd5482d6/torchvision/csrc/cuda/ROIPool_cuda.cu#L224

Just for my own sanity, I see that we first get the strides (to be passed into the kernel), and only then call ``grad.contiguous``, is this right? As I understand, making it contiguous should change the strides?

https://github.com/pytorch/vision/blob/dc3ac2909c2b9a5f6874ba1f1041ef49bd5482d6/torchvision/csrc/cuda/ROIPool_cuda.cu#L216-L224

**Comment by fmassa** [comment]
Yes, you are right, this is a bug! Removing the contiguous in this case should fix it. Can you also check if the same happens in ROIAlign, and send a PR? Thanks for the catch!

**Comment by lopuhin** [comment]
Thanks! Sure, will do

**Comment by fmassa** [comment]
And please also add a test for this case, if possible

**Comment by Naman-ntc** [comment]
Hi, thanks a lot for adding the PR for these ops.
Although I see that these ops haven't been documented neither in stable nor in master docs. Could someone add docs for these operations since it took me a lot of time to find about their existence and their signatures. I would have loved to do so, but I don't have enough knowledge about then to write docs.
Thanks

**Comment by varunagrawal** [comment]
@fmassa I guess we should rerun Sphinx and generate the docs afresh?

**Comment by fmassa** [comment]
@Naman-ntc I'm adding the doc entries for those functions now.

**Pull request merged**